capsules/extra/src/virtual_kv.rs

```diff
@@ -0,0 +1,349 @@
+    valid_ids: OptionalCell<StoragePermissions>,
+}
+
+impl<'a, V: kv_store::KV<'a>> ListNode<'a, VirtualKV<'a, V>> for VirtualKV<'a, V> {
+    fn next(&self) -> &'a ListLink<VirtualKV<'a, V>> {
+        &self.next
+    }
+}
+
+impl<'a, V: kv_store::KV<'a>> VirtualKV<'a, V> {
+    pub fn new(mux_kv: &'a MuxKV<'a, V>) -> VirtualKV<'a, V> {
+        Self {
+            mux_kv,
+            next: ListLink::empty(),
+            client: OptionalCell::empty(),
+            operation: OptionalCell::empty(),
+            unhashed_key: MapCell::empty(),
+            value: MapCell::empty(),
+            valid_ids: OptionalCell::empty(),
+        }
+    }
+
+    pub fn setup(&'a self) {
+        self.mux_kv.users.push_head(self);
+    }
+}
+
+impl<'a, V: kv_store::KV<'a>> kv_store::KV<'a> for VirtualKV<'a, V> {
+    fn set_client(&self, client: &'a dyn kv_store::StoreClient) {
+        self.client.set(client);
+    }
+
+    fn get(
+        &self,
+        key: LeasableMutableBuffer<'static, u8>,
+        value: LeasableMutableBuffer<'static, u8>,
+        permissions: StoragePermissions,
+    ) -> Result<
+        (),
+        (
+            LeasableMutableBuffer<'static, u8>,
+            LeasableMutableBuffer<'static, u8>,
+            Result<(), ErrorCode>,
+        ),
+    > {
+        if self.operation.is_some() {
+            return Err((key, value, Err(ErrorCode::BUSY)));
+        }
+
+        self.operation.set(Operation::Get);
+        self.valid_ids.set(permissions);
+        self.unhashed_key.replace(key);
+        self.value.replace(value);
+
+        self.mux_kv.do_next_op(false).map_err(|e| {
+            (
+                self.unhashed_key.take().unwrap(),
+                self.value.take().unwrap(),
+                e,
+            )
+        })
+    }
+
+    fn set(
+        &self,
+        key: LeasableMutableBuffer<'static, u8>,
+        value: LeasableMutableBuffer<'static, u8>,
+        permissions: StoragePermissions,
+    ) -> Result<
+        (),
+        (
+            LeasableMutableBuffer<'static, u8>,
+            LeasableMutableBuffer<'static, u8>,
+            Result<(), ErrorCode>,
+        ),
+    > {
+        match permissions.get_write_id() {
+            Some(_write_id) => {}
+            None => return Err((key, value, Err(ErrorCode::INVAL))),
+        }
+
+        if self.operation.is_some() {
+            return Err((key, value, Err(ErrorCode::BUSY)));
+        }
+
+        // The caller must ensure there is space for the header.
+        if value.len() < self.header_size() {
+            return Err((key, value, Err(ErrorCode::SIZE)));
+        }
+
+        // // Create the Tock header.
+        // let header = KeyHeader {
+        //     version: HEADER_VERSION,
+        //     length: (value.len() - HEADER_LENGTH) as u32,
+        //     write_id,
+        // };
+
+        // // Copy in the header to the buffer.
+        // header.copy_to_buf(value.as_slice());
```

Comment:
Is this intentionally left as comment?

Reply:
No that it is a mistake. I intend to fix.